[Improve][Transform-V2] Improve sql transform exception to locate error expression

misi1987107 opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist
  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/seatunnel/issues).
  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.
  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.
-->

### Purpose of this pull request

resolve issue https://github.com/apache/seatunnel/issues/9125


### Does this PR introduce _any_ user-facing change?

<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released SeaTunnel versions or within the unreleased branches such as dev.
If no, write 'No'.
If you are adding/modifying connector documents, please follow our new specifications: https://github.com/apache/seatunnel/issues/4544.
-->


### How was this patch tested?

<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If you are adding E2E test cases, maybe refer to https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql.conf, here is a good example.
-->


### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

Commit:
[feature] throw sql transform exception,locate error expression

Commit:
[feature] throw sql transform exception,locate error expression

Commit:
[improve] throw TransformException

Commit:
[improve] throw TransformException

Hisoka-X reviewed (commented):
Please update unit test.

Commit:
[improve] add test case

Commit:
[improve] add test case

Commit:
[improve] improve code

Commit:
[improve] improve code

Commit:
Merge branch 'dev' into bugfix-9125

Hisoka-X commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/exception/SqlTransformErrorCode.java:
```diff
@@ -0,0 +1,42 @@
+
+import org.apache.seatunnel.common.exception.SeaTunnelErrorCode;
+
+public enum SqlTransformErrorCode implements SeaTunnelErrorCode {
```
Please reuse https://github.com/apache/seatunnel/blob/37838165a8db066bf9f4b37baf3f0a5563788cc8/seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/exception/TransformCommonError.java#L33

Commit:
[improve] remove SqlTransformErrorCode\SqlTrasformException, reuse TransforCommonError

Commit:
[improve] remove SqlTransformErrorCode\SqlTrasformException, reuse TransforCommonError

Hisoka-X commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLEngine.java:
```diff
@@ -280,8 +282,17 @@ private Object[] project(Object[] inputFields) {
+                    fields[idx] = zetaSQLFunction.computeForValue(expression, inputFields);
+                    idx++;
+                } catch (Exception e) {
+                    if (e instanceof TransformException) {
+                        throw e;
```
why we should special care `TransformException`?

misi1987107 replied:
I previously thought that the computeForValue() method threw a TransformException(), so I should throw it outward.But I found this exception can be merged in the upper layer.

Hisoka-X commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/exception/TransformCommonError.java:
```diff
@@ -67,4 +68,11 @@ public static TransformException cannotFindInputTableError(String transform, Str
         return new TransformException(INPUT_TABLE_NOT_FOUND, params);
     }
+
+    public static TransformException sqlTransformError(String transform, String expression) {
```
```suggestion
    public static TransformException sqlExpressionError(String transform, String expression) {
```

misi1987107 replied:
get it

Hisoka-X commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/exception/TransformCommonErrorCode.java:
```diff
@@ -34,7 +34,10 @@ public enum TransformCommonErrorCode implements SeaTunnelErrorCode {
+            "The input table '<table>' of '<transform>' transform not found in upstream schema"),
+    EXPRESSION_EXECUTE_ERROR(
+            "TRANSFORM_COMMON-06",
+            "The expression '<expression>' of '<transform>' transform execute failed");
```
```suggestion
            "The expression '<expression>' of SQL transform execute failed");
```

The transform should always be SQL.

misi1987107 replied:
get it

Hisoka-X commented on seatunnel-transforms-v2/src/test/java/org/apache/seatunnel/transform/sql/SQLTransformTest.java:
```diff
@@ -441,4 +441,39 @@ public void tesBooleanField() {
+                        + "CAST(`FIELD1` AS STRING) AS FIELD1, "
+                        + "CAST(`FIELD1` AS decimal(22,4)) AS FIELD2, "
+                        + "CAST(`FIELD3` AS decimal(22,0)) AS FIELD3 "
+                        + "from dual";
```
I think we should handle `where` statement too.

Commit:
[improve] 1.add where statement test case
2.remove throw TransformException
3.rename sqlTransformError to sqlExpressionError
4.modify EXPRESSION_EXECUTE_ERROR description
5.add where statement processing

Commit:
[improve] 1.add where statement test case
2.remove throw TransformException
3.rename sqlTransformError to sqlExpressionError
4.modify EXPRESSION_EXECUTE_ERROR description
5.add where statement processing

Hisoka-X commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLEngine.java:
```diff
@@ -240,9 +241,13 @@ public List<SeaTunnelRow> transformBySQL(SeaTunnelRow inputRow, SeaTunnelRowType
+                return null;
+            }
+        } catch (Exception e) {
+            throw TransformCommonError.sqlWhereStatementError(selectBody.getWhere().toString());
```
We need add the exception into cause stack. Please refer https://github.com/apache/seatunnel/blob/454a88f81a3797aaec9decb820f876c89d599051/seatunnel-common/src/main/java/org/apache/seatunnel/common/exception/CommonError.java#L66

misi1987107 replied:
get it

Commit:
[improve] add the exception into cause stack

Commit:
[improve] add the exception into cause stack

Commit:
Merge remote-tracking branch 'origin/dev' into bugfix-9125

Commit:
[improve] modify throw the sqlTransformException

Commit:
[improve] modify throw the sqlTransformException

Hisoka-X commented on seatunnel-transforms-v2/src/test/java/org/apache/seatunnel/transform/sql/SQLTransformTest.java:
```diff
@@ -411,7 +411,7 @@ public void tesCastBooleanClausesWithField() {
                                 new SeaTunnelRow(new Object[] {Integer.valueOf(1), 0, "false333"}));
                     } catch (Exception e) {
                         Assertions.assertEquals(
-                                "ErrorCode:[COMMON-05], ErrorDescription:[Unsupported operation] - Unsupported CAST AS Boolean: false333",
```
The new exception looks like lose expression value in message. We should fix it.

misi1987107 replied:
already complete

Commit:
[improve] add validation of expression values in cause exception

Commit:
[improve] add validation of expression values in cause exception

hailin0 commented:
waiting for ci passed

misi1987107 commented:
@hailin0 The PR I submitted has nothing to do with this CI error , Can you help me take a look? I have tried again many times

Commit:
Merge branch 'dev' into bugfix-9125

Commit:
[improve] code format test

Commit:
[improve] code format test

Commit:
[improve] code format test

Commit:
[improve] code format test

Commit:
Merge remote-tracking branch 'origin/bugfix-9125' into bugfix-9125

Commit:
Merge branch 'dev' into bugfix-9125

Commit:
[improve] the expression exception is wraped
